panel/chat/step.py

```diff
@@ -0,0 +1,239 @@
+        self._title_pane = HTML(
+            self.param.title,
+            margin=0,
+            css_classes=["step-title"],
+        )
+        self.header = Row(
+            self._avatar_placeholder,
+            self._title_pane,
+            stylesheets=self._stylesheets + self.param.stylesheets.rx(),
+            css_classes=["step-header"],
+        )
+    def __enter__(self):
+        self.status = "running"
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if exc_type is not None:
+            self.status = "failed"
+            if self.failed_title is None:
+                # convert to str to wrap repr in apostrophes
+                self.failed_title = f"Failed due to: {str(exc_value)!r}"
+            raise exc_value
+        self.status = "success"
+
+    @param.depends("status", "default_avatars", watch=True)
+    def _render_avatar(self):
+        """
+        Render the avatar pane as some HTML text or Image pane.
+        """
+        extra_keys = set(self.default_avatars.keys()) - set(DEFAULT_STATUS_AVATARS.keys())
+        if extra_keys:
+            raise ValueError(
+                f"Invalid status avatars. Must be one of 'pending', 'running', 'success', 'failed'; "
+                f"got {extra_keys}."
+            )
+
+        avatar = avatar_lookup(
+            self.status,
+            None,
+            self.default_avatars,
+            DEFAULT_STATUS_AVATARS,
+        )
+        avatar_pane = build_avatar_pane(avatar, ["step-avatar"])
+        self._avatar_placeholder.update(avatar_pane)
+
+    @param.depends(
+        "status",
+        "default_title",
+        "pending_title",
+        "running_title",
+        "success_title",
+        "failed_title",
+        watch=True,
+        on_init=True,
+    )
+    def _update_title_on_status(self):
+        if self.status == "pending" and self.pending_title is not None:
+            title = self.pending_title
+        elif self.status == "running" and self.running_title is not None:
+            title = self.running_title
+        elif self.status == "success" and self.success_title is not None:
+            title = self.success_title
+        elif self.status == "failed" and self.failed_title is not None:
+            title = self.failed_title
+        else:
+            title = self.default_title
+        with param.edit_constant(self):
+            self.title = title
+
+    @param.depends("status", "collapsed_on_success", watch=True)
+    def _update_collapsed(self):
+        if self.status == "success" and self.collapsed_on_success:
+            self.collapsed = True
+
+    def stream_title(self, token: str, replace: bool = False):
+        """
+        Stream a token to the title header.
+
+        Arguments:
+        ---------
+        token : str
+            The token to stream.
+        replace : bool
+            Whether to replace the existing text.
+        """
+        if replace:
+            self.title = token
```

Comment:
Isn't title constant?!

Reply:
Also a little confused, I thought the title was always generated from the `<status>_title` parameters? Maybe you can only stream to the `running_title`?

Reply:
Or allow setting `status`?